src/session_state.rs

```diff
@@ -107,8 +103,15 @@ impl<Clock: embedded_time::Clock, const MSG_SIZE: usize, const MSG_COUNT: usize>
     }
 
     /// Called whenever an active connection has been made with a broker.
-    pub fn register_connection(&mut self) {
+    pub fn register_connection(&mut self, now: Instant<Clock>) {
         self.active = true;
+        self.ping_timeout = None;
+
+        // If the keep-alive interval is specified, set up a ping to go out before the interval
+        // elapses.
+        if let Some(interval) = self.keep_alive_interval {
+            self.next_ping.replace(now + interval - 500.milliseconds());
```

Comment:
I would go for half the interval (i.e. multiplicative), not some offset (additive) on it. That's how these heartbeat things are generally done and it's important that the timing scales multiplicatively with the timeout.

Reply:
From earlier:

> Took me a while to understand 500.milliseconds() * interval.0. What's the units of the factors and the units of the product?
> 500.milliseconds() * (interval/1.seconds()) would be meaningful to me (if that is possible).

An `embedded_time::Duration` can only be divided by integers, so it would just become `interval / 1`. There's no way to semantically "remove" the type by making it ratiometric here. The goal of `interval.0` * 500.milliseconds() is to take the internal "number" of seconds (a u32) and multiply that by 500ms. I'll write up docs around this.


> Same in L101.
> And who guarantees that the comparison is done in milliseconds. Are now and last always in milliseconds?

The comparison is not done in milliseconds. The comparison is always done in `Instant`'s. Adding a `Duration` to an `Instant` produces a new `Instant` with an offset of `Duration`.

Reply:
But there is no `interval.0 * 500.milliseconds()` anymore.

`now + interval / 2` doesn't work?

Reply:
The correct way would be to do `(interval.0 * 1000).milliseconds() / 2`, but yeah, that should also work